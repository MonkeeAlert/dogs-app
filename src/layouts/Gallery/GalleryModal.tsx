import {View, ImageBackground, StyleSheet} from 'react-native';
import React, {useEffect, useRef, useState} from 'react';
import {RouteProp, useRoute} from '@react-navigation/native';
import {Dimensions} from 'react-native';
import GoBack from '../../components/GoBack';
import {RootStackParamList} from '../Navigator/routes';
import {ListHeader} from './Components/ListHeader';
import {isAndroid} from '../../utils/functions';
import Animated, {
  Extrapolate,
  interpolate,
  useAnimatedGestureHandler,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from 'react-native-reanimated';
import {colors, springConfig, text} from '../../utils/constants';
import Api from '../../api/requests';
import {PanGestureHandler} from 'react-native-gesture-handler';
import SeeMore from './Components/SeeMore';
import CustomStatusBar from '../../components/CustomStatusBar';
import GalleryList from '../../components/lists/GalleryList';
import Loading from '../../components/Loading';

const FETCH_QUANTITY = 4;
const HEADER_HEIGHT = 67.33333587646484;
// const LOADING_STYLE = isAndroid() ? 'loadingAndroid' : 'loadingIOS';
const PLATFORM_BORDER = isAndroid() ? 28 : 14;

const GalleryModal = () => {
  const {params} = useRoute<RouteProp<RootStackParamList, 'Gallery'>>();

  const [data, setData] = useState<string[]>([]);
  const [size, setSize] = useState<{width: number; height: number}>({
    width: params.size.w,
    height: params.size.h,
  });

  const fetchDogs = useRef(async (s: string) => {
    try {
      const r = await Api.fetchDogBySubbreed(s, FETCH_QUANTITY);
      const d = [...data, ...(r.message as string[])];

      setData(d);
    } catch (err) {
      throw new Error(err as string);
    }
  });

  const panScrollY = useSharedValue(0);
  const scrollY = useSharedValue(0);

  const gestureHandler = useAnimatedGestureHandler(
    {
      onStart: (_, ctx: any) => {
        ctx.startY = panScrollY.value;
      },
      onActive: (evt, ctx) => {
        panScrollY.value = ctx.startY - evt.translationY;

        if (panScrollY.value <= 0) {
          panScrollY.value = 0;
        } else {
          if (panScrollY.value >= size.height) {
            panScrollY.value = size.height;
          } else {
            scrollY.value = 0;
          }
        }
      },
      onEnd: evt => {
        if (evt.translationY < 0) {
          panScrollY.value = withSpring(size.height, springConfig);
        } else {
          panScrollY.value = withSpring(0, springConfig);
        }
      },
    },
    [size.height],
  );

  const backgroundColor = useAnimatedStyle(
    () => ({
      opacity: interpolate(
        panScrollY.value,
        [0, size.height],
        [1, 0],
        Extrapolate.CLAMP,
      ),
    }),
    [size.height],
  );

  const goBackTransform = useAnimatedStyle(
    () => ({
      transform: [
        {
          translateX: interpolate(
            panScrollY.value,
            [0, size.height],
            [-100, 0],
            Extrapolate.CLAMP,
          ),
        },
      ],
    }),
    [size.height],
  );

  const panTransformStyle = useAnimatedStyle(() => {
    return {
      transform: [
        {
          translateY: interpolate(
            panScrollY.value,
            [0, size.height],
            [size.height - HEADER_HEIGHT, PLATFORM_BORDER],
            Extrapolate.CLAMP,
          ),
        },
      ],
    };
  }, [size.height]);

  // set image size
  useEffect(() => {
    setSize({width: params.size.w, height: params.size.h});

    if (params.isConnected && params.search) {
      fetchDogs.current(params.search);
    }

    return () => {
      panScrollY.value = withSpring(0, springConfig);
      scrollY.value = withSpring(0, springConfig);
      setData([]);
    };
  }, [params.uri]);

  return (
    <View style={styles.container}>
      <CustomStatusBar
        backgroundColor={'transparent'}
        barStyle={'light-content'}
      />

      <View style={styles.header}>
        <View style={isAndroid() ? styles.headerButtons : null}>
          <GoBack />

          <View style={styles.goBackContainer}>
            <Animated.Text style={[styles.goBackText, goBackTransform]}>
              Go back
            </Animated.Text>
          </View>
        </View>

        <Animated.View
          style={[backgroundColor, styles.background]}
          pointerEvents={'none'}>
          <ImageBackground source={{uri: params.img}} style={{...size}} />
        </Animated.View>
      </View>

      <PanGestureHandler
        testID={'GalleryModal_GestureHandler'}
        onGestureEvent={gestureHandler}>
        <Animated.View style={[panTransformStyle, styles.panGestureStyle]}>
          {!params.isConnected ? (
            <View
              style={{
                borderRadius: 14,
                overflow: 'hidden',
              }}>
              <ListHeader uri={params.uri} />
              <View style={styles.loadingList}>
                <Loading size={'large'} />
              </View>
            </View>
          ) : (
            <GalleryList
              images={data}
              HeaderComponent={<ListHeader uri={params.uri} />}
              FooterComponent={
                data.length >= 4 ? (
                  <SeeMore search={params.search ?? ''} />
                ) : (
                  <View />
                )
              }
            />
          )}
        </Animated.View>
      </PanGestureHandler>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.white,
    height: Dimensions.get('screen').height,
  },
  header: {
    position: 'absolute',
  },
  headerButtons: {
    position: 'relative',
    top: 28,
  },
  background: {
    zIndex: -1,
  },
  goBackContainer: {
    backgroundColor: 'transparent',
    position: 'absolute',
    top: 18,
    left: 35 + 14 * 1.5,
    zIndex: 1,
    overflow: 'hidden',
  },
  goBackText: {
    fontSize: text.m,
    fontWeight: '900',
    color: colors.darkGray,
  },
  panGestureStyle: {
    flex: 1,
    // borderRadiusTopLeft: 14,
    // borderRadiusTopRight: 14,
    overflow: 'hidden',
  },
  loadingList: {
    height: Dimensions.get('screen').height / 2,
    backgroundColor: colors.white,
  },
});

export default GalleryModal;
