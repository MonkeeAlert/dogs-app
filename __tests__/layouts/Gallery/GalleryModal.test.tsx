import {useRoute} from '@react-navigation/native';
import {act, render} from '@testing-library/react-native';
import React from 'react';
import {Image} from 'react-native';
import {create} from 'react-test-renderer';
import GalleryModal from '../../../src/layouts/Gallery/GalleryModal';
import {useSelector} from 'react-redux';

let mockedPromise = Promise.resolve();
const LIST = [
  'https://images.dog.ceo/breeds/labrador/n02099712_5648.jpg',
  'https://images.dog.ceo/breeds/retriever-golden/n02099601_8005.jpg',
  'https://images.dog.ceo/breeds/pyrenees/n02111500_7483.jpg',
];

jest.mock('../../../src/api/requests', () => ({
  fetchDogBySubbreed: jest.fn().mockImplementation(() => mockedPromise),
}));

describe('GalleryModal', () => {
  beforeEach(() => {
    useSelector.mockReturnValue({
      bookmarks: LIST,
    });

    const getSizeMock = jest.spyOn(Image, 'getSize');
    getSizeMock.mockImplementation(() => {});
  });

  it('matches snapshot', () => {
    useRoute.mockReturnValue({
      params: {
        uri: LIST[0],
        search: 'foo',
        img: 'foo.jpg',
        size: {w: 100, h: 100},
        isConnected: true,
      },
    });

    const tree = create(<GalleryModal />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('successfully fetches list', async () => {
    useRoute.mockReturnValue({
      params: {
        uri: LIST[0],
        search: '',
        img: 'foo.jpg',
        size: {w: 100, h: 100},
        isConnected: true,
      },
    });

    mockedPromise = Promise.resolve({message: ['foo', 'bar']});
    const spy = jest.spyOn(React, 'useEffect');
    render(<GalleryModal />);

    await act(async () => {
      expect(spy).toHaveBeenCalled();
    });
  });

  it('is not fetching list when promise was rejected', async () => {
    useRoute.mockReturnValue({
      params: {
        uri: LIST[0],
        search: '',
        img: 'foo.jpg',
        size: {w: 100, h: 100},
        isConnected: false,
      },
    });

    mockedPromise = Promise.reject('some error');
    const spy = jest.spyOn(React, 'useEffect');

    render(<GalleryModal />);
    await act(async () => {
      expect(spy).toHaveBeenCalled();
    });
  });

  it('is not fetching list when promise was rejected', async () => {
    useRoute.mockReturnValue({
      params: {
        uri: LIST[0],
        search: '',
        img: 'foo.jpg',
        size: {w: 100, h: 100},
        isConnected: true,
      },
    });

    mockedPromise = Promise.resolve({message: ['foo', 'bar']});
    const spy = jest.spyOn(React, 'useEffect');

    render(<GalleryModal />);
    await act(async () => {
      expect(spy).toHaveBeenCalled();
    });
  });
});
