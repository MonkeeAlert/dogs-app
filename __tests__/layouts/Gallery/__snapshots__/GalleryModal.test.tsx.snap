// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryModal matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fffafb",
      "height": 1334,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 44,
      }
    }
  />
  <View
    style={
      Object {
        "position": "absolute",
      }
    }
  >
    <View
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fffafb",
            "borderRadius": 35,
            "elevation": 1,
            "height": 35,
            "justifyContent": "center",
            "left": 14,
            "position": "absolute",
            "top": 14,
            "width": 35,
            "zIndex": 1,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="GoBack_Button"
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#2b2c2c",
                        "fontSize": 18,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  ï„¤
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "left": 56,
            "overflow": "hidden",
            "position": "absolute",
            "top": 18,
            "zIndex": 1,
          }
        }
      >
        <Text
          animatedStyle={
            Object {
              "value": Object {
                "transform": Array [
                  Object {
                    "translateX": -100,
                  },
                ],
              },
            }
          }
          collapsable={false}
          style={
            Object {
              "color": "#2b2c2c",
              "fontSize": 21,
              "fontWeight": "900",
              "transform": Array [
                Object {
                  "translateX": -100,
                },
              ],
            }
          }
        >
          Go back
        </Text>
      </View>
    </View>
    <View
      animatedStyle={
        Object {
          "value": Object {
            "opacity": 1,
          },
        }
      }
      collapsable={false}
      pointerEvents="none"
      style={
        Object {
          "opacity": 1,
          "zIndex": -1,
        }
      }
    >
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Object {
            "height": 100,
            "width": 100,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "foo.jpg",
            }
          }
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": 100,
                "width": 100,
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </View>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "transform": Array [
            Object {
              "translateY": 32.666664123535156,
            },
          ],
        },
      }
    }
    collapsable={false}
    forwardedRef={[Function]}
    handlerTag={1}
    handlerType="PanGestureHandler"
    onGestureHandlerEvent={[Function]}
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
        "transform": Array [
          Object {
            "translateY": 32.666664123535156,
          },
        ],
      }
    }
    testID="GalleryModal_GestureHandler"
  >
    <RCTScrollView
      ListFooterComponent={[Function]}
      ListHeaderComponent={
        <ListHeader
          uri="https://images.dog.ceo/breeds/labrador/n02099712_5648.jpg"
        />
      }
      bounces={false}
      contentContainerStyle={
        Array [
          Object {
            "backgroundColor": "#fffafb",
            "borderTopLeftRadius": 21,
            "borderTopRightRadius": 21,
            "paddingBottom": 28,
            "paddingHorizontal": 7,
            "paddingTop": 7,
          },
          Object {
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={16}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      testID="GalleryList"
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fffafb",
                "borderRadiusTopLeft": 14,
                "borderRadiusTopRight": 14,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingLeft": 14,
                "paddingRight": 11.5,
                "paddingVertical": 7,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2b2c2c",
                  "fontSize": 21,
                  "fontWeight": "900",
                  "paddingVertical": 14,
                }
              }
            >
              More 
              labradors
            </Text>
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="DefaultButton_TouchableOpacity"
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "#2BD9A5",
                      "borderRadius": 7,
                      "margin": 3.5,
                      "padding": 10,
                    }
                  }
                  testID="DefaultButton_View"
                >
                  <View
                    style={
                      Object {
                        "overflow": "hidden",
                      }
                    }
                  >
                    <View
                      testID="ListHeader_Icon"
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "backgroundColor": "transparent",
                            "justifyContent": "center",
                          }
                        }
                      >
                        <Text
                          allowFontScaling={false}
                          selectable={false}
                          style={
                            Array [
                              Object {
                                "color": "#fffafb",
                                "fontSize": 18,
                              },
                              Object {
                                "backgroundColor": "transparent",
                              },
                              Object {
                                "fontFamily": "Ionicons",
                                "fontStyle": "normal",
                                "fontWeight": "normal",
                              },
                              Object {},
                            ]
                          }
                          testID="iconIcon"
                        >
                          ï†“
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="DefaultButton_TouchableOpacity"
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "#2BD9A5",
                      "borderRadius": 7,
                      "margin": 3.5,
                      "padding": 10,
                    }
                  }
                  testID="DefaultButton_View"
                >
                  <View
                    style={
                      Object {
                        "overflow": "hidden",
                      }
                    }
                  >
                    <View>
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "backgroundColor": "transparent",
                            "justifyContent": "center",
                          }
                        }
                      >
                        <Text
                          allowFontScaling={false}
                          selectable={false}
                          style={
                            Array [
                              Object {
                                "color": "#fffafb",
                                "fontSize": 18,
                              },
                              Object {
                                "backgroundColor": "transparent",
                              },
                              Object {
                                "fontFamily": "Ionicons",
                                "fontStyle": "normal",
                                "fontWeight": "normal",
                              },
                              Object {},
                            ]
                          }
                          testID="iconIcon"
                        >
                          ï”œ
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
        >
          <View />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
